File: Exp 7 Application-based Algorithms using Greedy Approach /job_scheduling.cpp
#include <iostream>
#include <algorithm>
using namespace std;
struct Job
{
    int deadline;
    int profit;
};
int main()
{
    int size;
    cout << "Enter total job count: ";
    cin >> size;
    Job jobs[size];
    int max;
    cout << "Enter job details: " << endl;
    for (int i = 0; i < size; i++)
    {
        cout << "- Enter job " << i + 1 << " deadline: ";
        cin >> jobs[i].deadline;
        cout << "  Enter job " << i + 1 << " profit: ";
        cin >> jobs[i].profit;
        if (i == 0)
        {
            max = jobs[i].deadline;
        }
        else
        {
            if (max < jobs[i].deadline)
            {
                max = jobs[i].deadline;
            }
        }
    }

    sort(jobs, jobs + size, [](const Job &a, Job &b)
        { return a.profit > b.profit; });

    int slots[max];
    Job selectedJobs[max];

    for (int i = 0; i < max; i++)
    {
        slots[i] = 0;
    }

    int total_profit = 0;
    for (int i = 0; i < size; i++)
    {
        int s = jobs[i].deadline;
        for (int k = s - 1; k >= 0; k--)
        {
            if (slots[k] == 0)
            {
                slots[k] = 1;
                total_profit += jobs[i].profit;
                selectedJobs[k] = jobs[i];
                break;
            }
        }
    }
    cout << "Total profit: " << total_profit << endl;
    cout << "Selected jobs: " << endl;
    for (int i = 0; i < max; i++)
    {
        if (selectedJobs[i].deadline > 0)
            cout << selectedJobs[i].profit << "(" << selectedJobs[i].deadline << ")" << endl;    
    }

    return 0;
}

File: Exp 7 Application-based Algorithms using Greedy Approach /job_scheduling.py
class Job:
    """Job class to store deadline and profit"""
    def __init__(self, deadline=0, profit=0):
        self.deadline = deadline
        self.profit = profit


def main():
    size = int(input("Enter total job count: "))
    jobs = []
    max_deadline = 0
    
    print("Enter job details:")
    for i in range(size):
        deadline = int(input(f"- Enter job {i + 1} deadline: "))
        profit = int(input(f"  Enter job {i + 1} profit: "))
        jobs.append(Job(deadline, profit))
        if i == 0:
            max_deadline = deadline
        else:
            if max_deadline < deadline:
                max_deadline = deadline
    
    # Sort by profit in descending order
    jobs.sort(key=lambda x: x.profit, reverse=True)
    
    slots = [0] * max_deadline
    selected_jobs = [Job()] * max_deadline
    
    total_profit = 0
    for i in range(size):
        s = jobs[i].deadline
        for k in range(s - 1, -1, -1):
            if slots[k] == 0:
                slots[k] = 1
                total_profit += jobs[i].profit
                selected_jobs[k] = jobs[i]
                break
    
    print(f"Total profit: {total_profit}")
    print("Selected jobs:")
    for i in range(max_deadline):
        if selected_jobs[i].deadline > 0:
            print(f"{selected_jobs[i].profit}({selected_jobs[i].deadline})")


if __name__ == "__main__":
    main()
